server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -218,6 +255,25 @@ synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request,
                 return handler;
             }
 
+            /**
+             * Adds recovery source handler.
+             */
+            synchronized void reestablishRecovery(ReestablishRecoveryRequest request, ActionListener<RecoveryResponse> listener) {
```

Comment:
I wonder what will happen if the original recovery request has run to completion before `reestablishRecovery` has been called. At that point, the `RecoverySourceHandler`  has been  removed and we'll get a `ResourceNotFoundException`, even though the recovery completed. The target shard will then completely reset and restart the recovery. Not a biggie (given that the retry should be fast, as it has a retention lease and will do an ops-based recovery), just interesting case to consider.

Reply:
I will consider this scenario in a potential follow-up cleanups. This PR is already pretty big, so I want to hold off right now.

Reply:
As I said, I'm fine keeping the behavior as is (as it should incrementally restart due to ops-based recovery), just interesting to reason through the behavior.